### PR TITLE
[Train] Add `enable_reproducibility`

### LINKED ISSUE #22410: [Feature] Add function to limit sources of randomness
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Add a utility function that seeds everything and causes Train to avoid non-deterministic algorithms. See https://pytorch.org/docs/stable/notes/randomness.html.

### Use case

This is useful for testing, debugging, and development.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

This PR adds a feature that allows user to make their training runs more reproducible. I've implemented this feature by following PyTorch's guide on how to limit sources of randomness (https://pytorch.org/docs/stable/notes/randomness.html).

These changes will make it easier for us to benchmark Ray Train, and also make it easier for users to reproduce their experiments.

## Related issue number

Closes #22410

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch.py [hunk 1]
```diff
@@ -425,6 +468,19 @@ def accelerate() -> None:
         )
 
 
+@PublicAPI(stability="beta")
+def make_reproducible(seed: int = 0) -> None:
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply python/ray/train/torch.py]: Fixed

File: python/ray/train/torch.py [hunk 2]
```diff
@@ -132,6 +137,25 @@ def with_sampler(loader):
...
+                        worker_init_fn(worker_id)
+
+                    return wrapper
+
+                worker_init_fn = loader.worker_init_fn
+                generator = loader.generator
+                if self._is_seeded:
+                    worker_init_fn = seeded_worker_init_fn(loader.worker_init_fn)
+                    generator = torch.Generator()
+                    generator.manual_seed(0)
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply python/ray/train/torch.py]: Fixed

File: doc/source/train/api.rst [hunk 3]
```diff
@@ -183,6 +183,11 @@ train.torch.get_device
 
 .. autofunction:: ray.train.torch.get_device
 
+train.torch.make_reproducible
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/train/api.rst]: Fixed

File: doc/source/train/user_guide.rst [hunk 4]
```diff
@@ -977,6 +977,30 @@ the disk that from which your script was executed from.
     # View the PyTorch Profiler traces.
     $ open http://localhost:6006/#pytorch_profiler
 
+
```
Comment (amogkam) [inline doc/source/train/user_guide.rst:980]: Can you add a section bookmark above this? This would help if we want to easily reference this section from other parts of the docs.

File: python/ray/train/torch.py [hunk 5]
```diff
@@ -132,6 +137,25 @@ def with_sampler(loader):
                 # SequentialSampler.
                 shuffle = not isinstance(loader.sampler, SequentialSampler)
 
+                def seed_worker(worker_id):
+                    worker_seed = torch.initial_seed() % 2 ** 32
```
Comment (amogkam) [inline python/ray/train/torch.py:141]: oh this is interesting, out of curiosity, why do the % 2 **32?

File: python/ray/train/torch.py [hunk 6]
```diff
@@ -132,6 +137,25 @@ def with_sampler(loader):
                 # SequentialSampler.
                 shuffle = not isinstance(loader.sampler, SequentialSampler)
 
+                def seed_worker(worker_id):
```
Comment (amogkam) [inline python/ray/train/torch.py:140]: doesn't seem like worker_id is being used? Does it need to take in the arg?

**Review by amogkam — APPROVED** [review]
This is awesome! LGTM, just left some minor comments.

File: python/ray/train/torch.py (hunk 5 above)
Comment (bveeramani) [reply python/ray/train/torch.py:141]: I think it's because the `np.random.seed` requires a seed that is "convertible to 32 bit unsigned integers."

File: doc/source/train/user_guide.rst (hunk 4 above)
Comment (bveeramani) [reply doc/source/train/user_guide.rst:981]: Fixed

File: python/ray/train/torch.py (hunk 6 above)
Comment (bveeramani) [reply python/ray/train/torch.py]: Fixed
